### PR TITLE
SPOI-6189 Works with visualization now

### PR DESCRIPTION
**Pull request by amberarrow**
Top N WordCount app that monitors an input directory for text files; writes both per-file and global top N words and frequencies to snapshot server to support visualization.

### CODE REVIEW

**Comment by ilooner** [comment]
Make a ticket in MLHR for this and include that ticket number in the commit

File: demos/wordcount/src/main/java/com/datatorrent/demos/wordcount/FileWordCount.java [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+// control port:
+//   receive EOF tuples; after an EOF is received, emits tuple at the next endWindow
+// output port:
+//   Emit a singleton list containing a map with two entries: {fileName => L1, 'global' => L2},
+//   where L1/L2 are a list of topN pairs (word, frequency) for the current file and globally.
+//   The list wrapper is needed by AppDataSnapshotServerMap which is one of the receivers.
+//
+// Since the EOF is received by a single operator, this operator cannot be partitionable
+//@OperatorAnnotation(partitionable = true)
+public class FileWordCount extends BaseOperator
```
Comment (ilooner) [inline demos/wordcount/src/main/java/com/datatorrent/demos/wordcount/FileWordCount.java:55]: Use javadoc comment here. Remove the commented out OperatorAnnotation.

File: demos/wordcount/src/main/java/com/datatorrent/demos/wordcount/FileWordCount.java [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+  {
+    @Override
+    public void process(String msg)
+    {
+      if (msg.isEmpty()) {    // sanity check
+        throw new RuntimeException("Empty file path");
+      }
+      LOG.info("FileWordCount: EOF for {}, topN = {}", msg, topN);
+      fileName = msg;
+      eof = true;
```
Comment (ilooner) [inline demos/wordcount/src/main/java/com/datatorrent/demos/wordcount/FileWordCount.java:132]: Multiple files could be processed in a window and multiple eof messages could be sent.

File: demos/wordcount/src/main/java/com/datatorrent/demos/wordcount/FileWordCount.java [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+        fileFinalList.clear();
+        resultFileFinal.put(fileName, fileFinalList);
+        fileOutput.emit(resultFileFinal);
+
+        // reset for next file
+        eof = false;
+        fileName = null;
+        resultFileFinal.clear();
+      }
+      LOG.info("FileWordCount: endWindow for {}, no words, topN = {}", fileName, topN);
```
Comment (ilooner) [inline demos/wordcount/src/main/java/com/datatorrent/demos/wordcount/FileWordCount.java:194]: Multiple files could be processed in a window

File: demos/wordcount/src/main/java/com/datatorrent/demos/wordcount/FileWordCount.java [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+  // last component of path (i.e. only file name)
+  // incoming value from control tuple
+  protected String fileName;
+
+  // wordMapFile   : {word => frequency} map, current file, all words
+  // wordMapGlobal : {word => frequency} map, global, all words
+  //
+  protected Map<String, WCPair>
+    wordMapFile = new HashMap<>(),
+    wordMapGlobal = new HashMap<>();
```
Comment (ilooner) [inline demos/wordcount/src/main/java/com/datatorrent/demos/wordcount/FileWordCount.java]: do our style rules allow this?

File: demos/wordcount/src/main/java/com/datatorrent/demos/wordcount/LineReader.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
+package com.datatorrent.demos.wordcount;
```
Comment (ilooner) [inline demos/wordcount/src/main/java/com/datatorrent/demos/wordcount/LineReader.java:16]: Missing license header

File: demos/wordcount/src/main/java/com/datatorrent/demos/wordcount/WordCountWriter.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
+package com.datatorrent.demos.wordcount;
```
Comment (ilooner) [inline demos/wordcount/src/main/java/com/datatorrent/demos/wordcount/WordCountWriter.java:16]: Missing license header

File: demos/wordcount/src/main/java/com/datatorrent/demos/wordcount/WordReader.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
+package com.datatorrent.demos.wordcount;
```
Comment (ilooner) [inline demos/wordcount/src/main/java/com/datatorrent/demos/wordcount/WordReader.java:16]: license header

File: demos/wordcount/src/main/java/com/datatorrent/demos/wordcount/WindowWordCount.java [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+import java.util.Map;
+
+//import org.apache.commons.lang.mutable.MutableInt;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.datatorrent.api.DefaultInputPort;
+import com.datatorrent.api.DefaultOutputPort;
+//import com.datatorrent.api.annotation.OperatorAnnotation;
```
Comment (ilooner) [inline demos/wordcount/src/main/java/com/datatorrent/demos/wordcount/WindowWordCount.java]: delete commented import

File: demos/wordcount/src/main/java/com/datatorrent/demos/wordcount/WindowWordCount.java [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+import com.datatorrent.api.Context.OperatorContext;
+import com.datatorrent.common.util.BaseOperator;
+//import com.datatorrent.lib.util.BaseUniqueKeyCounter;
```
Comment (ilooner) [inline demos/wordcount/src/main/java/com/datatorrent/demos/wordcount/WindowWordCount.java]: delete commented import

File: demos/wordcount/src/main/java/com/datatorrent/demos/wordcount/FileWordCount.java [hunk 10]
```diff
@@ -0,0 +1,282 @@
...
+package com.datatorrent.demos.wordcount;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+//import org.apache.commons.lang.mutable.MutableInt;
```
Comment (ilooner) [inline demos/wordcount/src/main/java/com/datatorrent/demos/wordcount/FileWordCount.java:25]: delete commented import

**Comment by amberarrow** [comment]
I think I've addressed all the comments so far in commit 0a5914ce8066

**Pull request merged**